ulib/LowStar.Vector.fst

```diff
@@ -0,0 +1,558 @@
+  slice_append (S.upd (as_seq hh0 vec) (U32.v i) v) 0 (U32.v i) (U32.v i + 1);
+  slice_append (S.upd (as_seq hh0 vec) (U32.v i) v) 0 (U32.v i + 1) (U32.v (size_of vec))
+
+private val resize_ratio: uint32_t
+private let resize_ratio = 2ul
+
+private val new_capacity: cap:uint32_t{cap > 0ul} -> Tot uint32_t
+private let new_capacity cap =
+  if cap >= max_uint32 / resize_ratio then max_uint32
+  else cap * resize_ratio
+
+val insert: 
+  #a:Type -> vec:vector a -> v:a -> 
+  HST.ST (vector a)
+    (requires (fun h0 -> 
+      live h0 vec /\ freeable vec /\ not (is_full vec) /\
+      HST.is_eternal_region (frameOf vec)))
+    (ensures (fun h0 nvec h1 ->
+      frameOf vec = frameOf nvec /\
+      hmap_dom_eq h0 h1 /\
+      live h1 nvec /\ freeable nvec /\
+      modifies (loc_union (loc_addr_of_vector vec) 
+      			  (loc_vector nvec)) h0 h1 /\
+      size_of nvec = size_of vec + 1ul /\
+      get h1 nvec (size_of vec) == v /\
+      S.equal (as_seq h1 nvec) (S.snoc (as_seq h0 vec) v)))
+#reset-options "--z3rlimit 20"
+let insert #a vec v =
+  let sz = Vec?.sz vec in
+  let cap = Vec?.cap vec in
+  let vs = Vec?.vs vec in
+  if sz = cap 
+  then (let ncap = new_capacity cap in
+       let nvs = B.malloc (B.frameOf vs) v ncap in
+       B.blit vs 0ul nvs 0ul sz;
+       B.upd nvs sz v;
+       B.free vs;
+       Vec (sz + 1ul) ncap nvs)
+  else
+    (B.upd vs sz v;
+    Vec (sz + 1ul) cap vs)
+
+val flush:
+  #a:Type -> vec:vector a -> ia:a ->
+  i:uint32_t{i <= size_of vec} ->
+  HST.ST (vector a)
+    (requires (fun h0 ->
+      live h0 vec /\ freeable vec /\
+      HST.is_eternal_region (frameOf vec)))
+    (ensures (fun h0 fvec h1 ->
+      frameOf vec = frameOf fvec /\
+      hmap_dom_eq h0 h1 /\
+      live h1 fvec /\ freeable fvec /\
+      modifies (loc_union (loc_addr_of_vector vec) 
+      			  (loc_vector fvec)) h0 h1 /\
+      size_of fvec = size_of vec - i /\
+      S.equal (as_seq h1 fvec) 
+	      (S.slice (as_seq h0 vec) (U32.v i) (U32.v (size_of vec)))))
+let flush #a vec ia i =
+  let fsz = Vec?.sz vec - i in
+  let asz = if Vec?.sz vec = i then 1ul else fsz in
+  let vs = Vec?.vs vec in
+  let fvs = B.malloc (B.frameOf vs) ia asz in
+  B.blit vs i fvs 0ul fsz;
+  B.free vs;
+  Vec fsz asz fvs
+
+/// Iteration
+
+val fold_left_seq:
+  #a:Type -> #b:Type0 -> seq:S.seq a ->
+  f:(b -> a -> GTot b) -> ib:b -> 
+  GTot b (decreases (S.length seq))
+let rec fold_left_seq #a #b seq f ib =
+  if S.length seq = 0 then ib
+  else fold_left_seq (S.tail seq) f (f ib (S.head seq))
+
+val fold_left_buffer:
+  #a:Type -> #b:Type0 -> len:uint32_t ->
+  buf:B.buffer a{B.len buf = len} ->
+  f:(b -> a -> Tot b) -> ib:b ->
+  HST.ST b
+    (requires (fun h0 -> B.live h0 buf))
+    (ensures (fun h0 v h1 -> 
+      h0 == h1 /\
+      v == fold_left_seq (B.as_seq h0 buf) f ib))
+    (decreases (B.length buf))
+let rec fold_left_buffer #a #b len buf f ib =
+  if len = 0ul then ib
+  else (fold_left_buffer (len - 1ul) (B.sub buf 1ul (len - 1ul)) 
```

Comment:
This one could conceivably use the `C.Loops` module for a more efficient iteration, but that can come in a later pull request

Reply:
(especially since we need to move C.Loops to LowStar.Loops first)